plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,654 @@
+					    MBOX_ACCESS, /* value */
+					    mbox_reg, /* index */
+					    (guchar *) buf, /* data */
+					    4,  /* length */
+					    NULL, /* actual length */
+					    MBOX_TIMEOUT,
+					    NULL, error)) {
+		g_prefix_error (error, "failed to get mmio 0x%x: ", mbox_reg);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_set_mmio (FuDevice *device, guint16 mbox_reg, guchar *buf,
+			        GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+
+	if (!g_usb_device_control_transfer (usb_device,
+					    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    REQ_HUB_SET_MMIO, /* request */
+					    MBOX_ACCESS,      /* value */
+					    mbox_reg,	      /* index */
+					    (guchar *) buf,   /* data */
+					    4,		      /* length */
+					    NULL,	      /* actual length */
+					    MBOX_TIMEOUT,
+					    NULL, error)) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+			     "failed to set mmio 0x%x: ", mbox_reg);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/*
+ * Read up to 64 bytes of data from the mbox data registers to a buffer.
+ * The mailbox can hold 64 bytes of data in 16 doubleword data registers.
+ * To get data from NVM or DROM to mbox registes issue a NVM Read or DROM
+ * read operation before reading the mbox data registes.
+ */
+static gboolean
+fu_dell_dock_usb4_mbox_data_read (FuDevice *device, guchar *data, guint8 length,
+				  GError **error)
+{
+	guchar *ptr = data;
+	gint	i;
+
+	if (length > 64 || length % 4) {
+		g_prefix_error (error, "invalid firmware data read length %u",
+				length);
+		return FALSE; /* FIXME check return code */
+	}
+	/* read 4 bytes per iteration */
+	for (i = 0; i < length / 4; i++) {
+		if (!fu_dell_dock_usb4_hub_get_mmio(device, i, ptr, error))
+			return FALSE; /* FIXME also print a error message */
+		ptr += 4;
+	}
+	return TRUE;
+}
+
+/*
+ * The mailbox can hold 64 bytes in 16 doubleword data registers.
+ * A NVM write operation writes data from these registers to NVM
+ * at the set offset
+ */
+static gboolean
+fu_dell_dock_usb4_mbox_data_write (FuDevice *device, const guchar *data,
+				   guint8 length, GError **error)
+{
+	guchar *ptr = (guchar *) data;
+
+	if (length > 64 || length % 4) {
+		g_prefix_error (error,
+				"invalid firmware data write length %u\n",
+				length);
+		return FALSE; /* FIXME check return code */
+	}
+	/* writes 4 bytes per iteration */
+	for (gint i = 0; i < length / 4; i++) {
+		if (!fu_dell_dock_usb4_hub_set_mmio (device, i, ptr, error))
+			return FALSE;
+		ptr += 4;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_operation (FuDevice *device, guint16 opcode,
+				 guchar *metadata, guint8 *status,
+				 GError **error)
+{
+	struct mbox_regx *regx;
+	guchar buf[4];
+	gint   i;
+
+	regx = (struct mbox_regx *)buf;
+	memcpy(&regx->opcode, &opcode, sizeof(regx->opcode));
```

Comment:
@superm1 the sample code indicated to sort the bytes in `little endian`, here I use `memcpy` because my laptop is default `LE`. Is there a more recommended way to do this?
 
```
regx->opcode = libusb_cpu_to_le16(opcode);
```

Reply:
Yes, there's a bunch of byte order macros in glib you can use: https://developer.gnome.org/glib/stable/glib-Byte-Order-Macros.html